File: LeetCode/remote/first_bad_version.cs
using System;
using System.Collections.Generic;

// https://leetcode.com/submissions/detail/39112846/
//
// Submission Details
// 21 / 21 test cases passed.
//  Status: Accepted
//  Runtime: 72 ms
//
class Program {
    private int count = 0;
    private readonly int badVersion;
    private Dictionary<int, bool> cache = new Dictionary<int, bool>();

    public Program(int badVer) {
        badVersion = badVer;
    }

    private bool IsBadVersion(int version) {
        count++;
        if (version >= badVersion) {
            return true;
        }

        return false;
    }

    private bool CachedBadVersion(int version) {
        if (!cache.ContainsKey(version)) {
            cache[version] = IsBadVersion(version);
        }

        return cache[version];
    }

    public int FirstBadVersion(int n) {
        int low = 1;
        int high = n;
        if (low == 1 && CachedBadVersion(low)) {
            return 1;
        }

        while (low < high) {
            int mid = low + (high - low) / 2;
            if (!CachedBadVersion(mid) &&
                mid < n &&
                CachedBadVersion(mid + 1)) {
                return mid + 1;
            }
            
            if (CachedBadVersion(mid)) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }

        return -1;
    }

    public int GetCount() {
        return count;
    }

    static void Main() {
        Program p = new Program(5);
        Console.WriteLine("{0} - {1}", p.FirstBadVersion(101), p.GetCount());
        p = new Program(1);
        Console.WriteLine("{0} - {1}", p.FirstBadVersion(1), p.GetCount());
        p = new Program(10);
        Console.WriteLine("{0} - {1}", p.FirstBadVersion(9), p.GetCount());
    }
}

File: LeetCode/remote/first_bad_version.rb
#   https://leetcode.com/problems/first-bad-version/
#
#   You are a product manager and currently leading a team to develop a new product.
#   Unfortunately, the latest version of your product fails the quality check.
#   Since each version is developed based on the previous version, all the versions after a bad version are also bad. 

#   
#   Submission Details
#   21 / 21 test cases passed.
#       Status: Accepted
#       Runtime: 64 ms
#           
#           Submitted: 0 minutes ago
#
#   https://leetcode.com/submissions/detail/53263348/

def first_bad_version(n)
    low = 1
    high = n
    while low <= high
        mid = low + (high - low) / 2
        if is_bad_version(mid)
            high = mid - 1
        else
            low = mid + 1
        end
    end
    
    return low
end

def first_bad_version(n)
    low, high = 1, n
    while low < high
        mid = low + ((high - low) >> 2)
        if is_bad_version(mid)
           high = mid 
       else
           low = mid + 1
       end
    end
    
    low
end
